README.md
# Simple Card Brand

Determine a payment card's brand from the leading digits and length of the account number (PAN).

Uses the rules documented on [Wikipedia](https://en.wikipedia.org/wiki/Payment_card_number).

## Card Brands

- American Express (:americanexpress)
- BORICA (:borica)
- China T-Union (:chinatunion)
- China UnionPay (:chinaunionpay)
- Dankort (:dankort)
- Diners Club (:dinersclub)
- Diners Club International (:dinersclubinternational)
- Discover (:discover)
- GPN (:gpn)
- Humo (:humo)
- InstaPayment (:instapayment)
- InterPayment (:interpayment)
- JCB: (:jcb)
- LankaPay: (:lankapay)
- Maestro (:maestro)
- Maestro UK (:maestrouk)
- Mastercard (:mastercard)
- Mir (:mir)
- Napas (:napas)
- RuPay (:rupay)
- Troy (:troy)
- UATP (:uatp)
- UkrCard (:ukrcard)
- UzCard (:uzcard)
- Verve (:verve)
- Visa (:visa)
- Visa Electron (:visaelectron)

### Conflicts

The [Wikipedia](https://en.wikipedia.org/wiki/Payment_card_number) page has several account number range conflicts that need to be researched.

- 65 is always recognized as Discover.
  Overrides RuPay and Troy.

- 6759 is documented as both Maestro and Maestro UK.
  - Recognized as Maestro UK.

- 54 is documented as both Diners Club in the US and Mastercard.
  - Recognized as Diners Club.

## Mix Commands

- mix format
- mix test
- mix coveralls.html
- mix credo --strict
- mix dialyzer
- mix inch
  - Documentation coverage test.
  - Use as guidance, not pass/fail CI.
  - Fails to note documentation on new `defguard` macro.
- mix hex.outdated
- mix hex.audit
- mix deps.update XXX
- mix sobelow --config
  - Better suited for Phoenix apps. Included for future reference.

## Installation

TO DO

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `simplecardbrand` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:simplecardbrand, "~> 0.1.0"}
  ]
end
```

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at <https://hexdocs.pm/simplecardbrand>.

## Adopting Elixir

I'm using this project as a reference for various quality control tools.
As a result, there's a lot of code quality tools for such a small single-file module.
Using the [Adopting Elixir](https://www.kobo.com/us/en/ebook/adopting-elixir) recommendations.


CHANGELOG.md
# CHANGELOG

## 1.0.1 -- 2023-09-08

### Added

- Test guards.
- Test CLI module.

### Changed

- Remove code that maps PAN starting with 65 to :rupay.
  PAN starting with 65 is Discover.

- CI changes:
  - Elixir 1.14.5
  - OTP 25.3.2.5
  - Add Elixir 1.15.5
  - Add OTP 26.0.2

- Code Quality Changes (Adopting Elixir):
  - Credo config
  - Document dialyzer
  - InchEx
  - ExCoveralls
  - Sobelow

- Update dependencies:
  - `dialyxir`
  - `ex_doc`

## 1.0.0 -- 2023-08-08

Publish

### Added

- CLI for `card_brand/2`.
- Add GitHub workflow.

## 0.3.0 - 2023-08-08

### Added

- Add README and CHANGELOG to Hex docs.
- Add pre-commit tasks.
- Add command-line interface.

### Changed

- Return reason tuple on error.
  - `{:error, {:pan_too_short,"Minimum PAN length is 12, found 10."}}`
  - `{:error, {:pan_unknown,"Unknown card brand."}}`
  - `{:error, {:pan_too_long, "Maximum PAN length is 19, found 20."}}`
- Replace Erlang binary/bytes types with String.t in @spec's.

## 0.2.0 - 2023-08-07 WIP

### Added

- Brands: Dankort, GPN, LankaPay, Troy, UkrCard, Verve.
  - Brands are now complete.
- In-line document tests.
- Generate documentation.

### Changed

- Correct major error of `card_brand/1` passing CodePoints to `card_brand/2`.
- Identify PAN starting just 65 as Discover, vs. RuPay or Troy.
- Minor credo refactors.

## v0.1.0 - 2023-08-05 WIP

### Added

- Add guards to `card_brand`.
- Rough move to mix project, tests are functional.

## v0.0.3 - 2023-08-04 WIP

### Added

- Make `pan_length` first parameter of `pan_range`.
- China UnionPay, China T-Union, Diners Club, Diners Club International
- JCB, Maestro, Maestro UK, Visa Electron
- BORICA, Humo, InstaPayment, InterPayment, Mir, Napas, UATP, UzCard

## v0.0.2 - 2023-08-03 WIP

- Initial release.

### Added

- American Express, Discover, Mastercard, and Visa.


test/cli_test.exs
defmodule SimpleCardBrandCliTest do
  @moduledoc """
  Test the CLI.
  """

  use ExUnit.Case
  import ExUnit.CaptureIO

  test "Valid CLI Visa PAN." do
    assert capture_io(fn -> Commandline.CLI.main(["4111111111111111"]) end) ==
             "PAN: 4111111111111111 -> Brand: visa\n"
  end

  test "Valid CLI Mastercard PAN prefix and length." do
    assert capture_io(fn -> Commandline.CLI.main(["5123456", "--pan-length=16"]) end) ==
             "PAN: 5123456 -> Brand: mastercard\n"
  end

  test "Invalid PAN." do
    assert capture_io(fn -> Commandline.CLI.main(["411111111111111156784"]) end) ==
             "PAN: 411111111111111156784 -> Maximum PAN length is 19, found 21.\n"
  end

  test "Invalid PAN prefix and length." do
    assert capture_io(fn -> Commandline.CLI.main(["5123456", "--pan-length=12"]) end) ==
             "PAN: 5123456 -> Unknown card brand.\n"
  end
end


test/gpn_test.exs
defmodule SimpleCardBrandGPN do
  @moduledoc """
  GPN tests.
  """
  use ExUnit.Case

  test "GpnWikipedia" do
    for first <- ["1", "2", "6", "7", "8", "9"] do
      assert SimpleCardBrand.card_brand(first, 16) == {:ok, :gpn}
    end
  end
end


test/jcb_test.exs
defmodule SimpleCardBrandJCB do
  @moduledoc """
  JCB tests.
  """

  use ExUnit.Case, async: true
  use ExUnitProperties

  property "JCBWikipedia" do
    check all(
            fragment <- StreamData.integer(3528..3589),
            pan_length <- StreamData.integer(16..19)
          ) do
      # Skip RuPay brand range.
      if fragment not in 3530..3539 and fragment not in 3560..3569 do
        assert SimpleCardBrand.card_brand(Integer.to_string(fragment), pan_length) ==
                 {:ok, :jcb}
      end
    end
  end
end


mix.exs
defmodule SimpleCardBrand.MixProject do
  use Mix.Project

  @source_url "https://github.com/rgacote/SimpleCardBrand"
  @version "1.0.1"
  def project do
    [
      app: :simplecardbrand,
      version: @version,
      elixir: "~> 1.14",
      build_embedded: Mix.env() == :prod,
      escript: [main_module: Commandline.CLI],
      description: "Identify card brands",
      package: package(),
      deps: deps(),

      # Docs
      name: "SimpleCardBrand",
      docs: docs(),
      test_coverage: [tool: ExCoveralls, export: "cov"],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],
      test_coverage: [tool: ExCoveralls]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:credo, "~> 1.7", only: [:dev, :test], runtime: false},
      {:stream_data, "~> 0.6", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.30", only: :dev, runtime: false},
      {:dialyxir, "~> 1.4", only: [:dev], runtime: false},
      {:inch_ex, "~>  2.1.0-rc.1", only: [:dev, :test], runtime: false},
      {:excoveralls, "~> 0.17.1", only: [:test], runtime: false},
      {:sobelow, "~> 0.13", only: [:dev, :test], runtime: false}
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"}
    ]
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README*", "LICENSE*"],
      maintainers: ["Raymond GA Côté"],
      licenses: ["BSD-3-Clause"],
      links: %{"GitHub" => "https://github.com/rgacote/SimpleCardBrand"}
    ]
  end

  # Example
  # https://github.com/elixir-ecto/ecto/blob/b275ef25349b14ee0e01da7388dea544fecc1fc2/mix.exs
  defp docs do
    [
      main: "SimpleCardBrand",
      source_ref: "v#{@version}",
      source_url: @source_url,
      extra_section: "GUIDES",
      skip_undefined_reference_warnings_on: ["CHANGELOG.md"],
      extras: extras(),
      groups_for_extras: groups_for_extras()
    ]
  end

  defp extras() do
    ["README.md", "CHANGELOG.md"]
  end

  defp groups_for_extras do
    [Introduction: "README.md"]
  end
end


.credo.exs
# This file contains the configuration for Credo and you are probably reading
# this after creating it with `mix credo.gen.config`.
#
# If you find anything wrong or unclear in this file, please report an
# issue on GitHub: https://github.com/rrrene/credo/issues
#
%{
  #
  # You can have as many configs as you like in the `configs:` field.
  configs: [
    %{
      #
      # Run any config using `mix credo -C <name>`. If no config name is given
      # "default" is used.
      #
      name: "default",
      #
      # These are the files included in the analysis:
      files: %{
        #
        # You can give explicit globs or simply directories.
        # In the latter case `**/*.{ex,exs}` will be used.
        #
        included: [
          "lib/",
          "src/",
          "test/",
          "web/",
          "apps/*/lib/",
          "apps/*/src/",
          "apps/*/test/",
          "apps/*/web/"
        ],
        excluded: [~r"/_build/", ~r"/deps/", ~r"/node_modules/"]
      },
      #
      # Load and configure plugins here:
      #
      plugins: [],
      #
      # If you create your own checks, you must specify the source files for
      # them here, so they can be loaded by Credo before running the analysis.
      #
      requires: [],
      #
      # If you want to enforce a style guide and need a more traditional linting
      # experience, you can change `strict` to `true` below:
      #
      strict: false,
      #
      # To modify the timeout for parsing files, change this value:
      #
      parse_timeout: 5000,
      #
      # If you want to use uncolored output by default, you can change `color`
      # to `false` below:
      #
      color: true,
      #
      # You can customize the parameters of any check by adding a second element
      # to the tuple.
      #
      # To disable a check put `false` as second element:
      #
      #     {Credo.Check.Design.DuplicatedCode, false}
      #
      checks: %{
        enabled: [
          #
          ## Consistency Checks
          #
          {Credo.Check.Consistency.ExceptionNames, []},
          {Credo.Check.Consistency.LineEndings, []},
          {Credo.Check.Consistency.ParameterPatternMatching, []},
          {Credo.Check.Consistency.SpaceAroundOperators, []},
          {Credo.Check.Consistency.SpaceInParentheses, []},
          {Credo.Check.Consistency.TabsOrSpaces, []},

          #
          ## Design Checks
          #
          # You can customize the priority of any check
          # Priority values are: `low, normal, high, higher`
          #
          {Credo.Check.Design.AliasUsage,
           [priority: :low, if_nested_deeper_than: 2, if_called_more_often_than: 0]},
          # You can also customize the exit_status of each check.
          # If you don't want TODO comments to cause `mix credo` to fail, just
          # set this value to 0 (zero).
          #
          {Credo.Check.Design.TagTODO, [exit_status: 2]},
          {Credo.Check.Design.TagFIXME, []},

          #
          ## Readability Checks
          #
          {Credo.Check.Readability.AliasOrder, []},
          {Credo.Check.Readability.FunctionNames, []},
          {Credo.Check.Readability.LargeNumbers, []},
          {Credo.Check.Readability.MaxLineLength, [priority: :low, max_length: 120]},
          {Credo.Check.Readability.ModuleAttributeNames, []},
          {Credo.Check.Readability.ModuleDoc, []},
          {Credo.Check.Readability.ModuleNames, []},
          {Credo.Check.Readability.ParenthesesInCondition, []},
          {Credo.Check.Readability.ParenthesesOnZeroArityDefs, []},
          {Credo.Check.Readability.PipeIntoAnonymousFunctions, []},
          {Credo.Check.Readability.PredicateFunctionNames, []},
          {Credo.Check.Readability.PreferImplicitTry, []},
          {Credo.Check.Readability.RedundantBlankLines, []},
          {Credo.Check.Readability.Semicolons, []},
          {Credo.Check.Readability.SpaceAfterCommas, []},
          {Credo.Check.Readability.StringSigils, []},
          {Credo.Check.Readability.TrailingBlankLine, []},
          {Credo.Check.Readability.TrailingWhiteSpace, []},
          {Credo.Check.Readability.UnnecessaryAliasExpansion, []},
          {Credo.Check.Readability.VariableNames, []},
          {Credo.Check.Readability.WithSingleClause, []},

          #
          ## Refactoring Opportunities
          #
          {Credo.Check.Refactor.Apply, []},
          {Credo.Check.Refactor.CondStatements, []},
          {Credo.Check.Refactor.CyclomaticComplexity, []},
          {Credo.Check.Refactor.FunctionArity, []},
          {Credo.Check.Refactor.LongQuoteBlocks, []},
          {Credo.Check.Refactor.MatchInCondition, []},
          {Credo.Check.Refactor.MapJoin, []},
          {Credo.Check.Refactor.NegatedConditionsInUnless, []},
          {Credo.Check.Refactor.NegatedConditionsWithElse, []},
          {Credo.Check.Refactor.Nesting, []},
          {Credo.Check.Refactor.UnlessWithElse, []},
          {Credo.Check.Refactor.WithClauses, []},
          {Credo.Check.Refactor.FilterCount, []},
          {Credo.Check.Refactor.FilterFilter, []},
          {Credo.Check.Refactor.RejectReject, []},
          {Credo.Check.Refactor.RedundantWithClauseResult, []},

          #
          ## Warnings
          #
          {Credo.Check.Warning.ApplicationConfigInModuleAttribute, []},
          {Credo.Check.Warning.BoolOperationOnSameValues, []},
          {Credo.Check.Warning.Dbg, []},
          {Credo.Check.Warning.ExpensiveEmptyEnumCheck, []},
          {Credo.Check.Warning.IExPry, []},
          {Credo.Check.Warning.IoInspect, []},
          {Credo.Check.Warning.MissedMetadataKeyInLoggerConfig, []},
          {Credo.Check.Warning.OperationOnSameValues, []},
          {Credo.Check.Warning.OperationWithConstantResult, []},
          {Credo.Check.Warning.RaiseInsideRescue, []},
          {Credo.Check.Warning.SpecWithStruct, []},
          {Credo.Check.Warning.WrongTestFileExtension, []},
          {Credo.Check.Warning.UnusedEnumOperation, []},
          {Credo.Check.Warning.UnusedFileOperation, []},
          {Credo.Check.Warning.UnusedKeywordOperation, []},
          {Credo.Check.Warning.UnusedListOperation, []},
          {Credo.Check.Warning.UnusedPathOperation, []},
          {Credo.Check.Warning.UnusedRegexOperation, []},
          {Credo.Check.Warning.UnusedStringOperation, []},
          {Credo.Check.Warning.UnusedTupleOperation, []},
          {Credo.Check.Warning.UnsafeExec, []}
        ],
        disabled: [
          #
          # Checks scheduled for next check update (opt-in for now, just replace `false` with `[]`)

          #
          # Controversial and experimental checks (opt-in, just move the check to `:enabled`
          #   and be sure to use `mix credo --strict` to see low priority checks)
          #
          {Credo.Check.Consistency.MultiAliasImportRequireUse, []},
          {Credo.Check.Consistency.UnusedVariableNames, []},
          {Credo.Check.Design.DuplicatedCode, []},
          {Credo.Check.Design.SkipTestWithoutComment, []},
          {Credo.Check.Readability.AliasAs, []},
          {Credo.Check.Readability.BlockPipe, []},
          {Credo.Check.Readability.ImplTrue, []},
          {Credo.Check.Readability.MultiAlias, []},
          {Credo.Check.Readability.NestedFunctionCalls, []},
          {Credo.Check.Readability.OneArityFunctionInPipe, []},
          {Credo.Check.Readability.SeparateAliasRequire, []},
          {Credo.Check.Readability.SingleFunctionToBlockPipe, []},
          {Credo.Check.Readability.SinglePipe, []},
          {Credo.Check.Readability.Specs, []},
          {Credo.Check.Readability.StrictModuleLayout, []},
          {Credo.Check.Readability.WithCustomTaggedTuple, []},
          {Credo.Check.Readability.OnePipePerLine, []},
          {Credo.Check.Refactor.ABCSize, []},
          {Credo.Check.Refactor.AppendSingleItem, []},
          {Credo.Check.Refactor.DoubleBooleanNegation, []},
          {Credo.Check.Refactor.FilterReject, []},
          {Credo.Check.Refactor.IoPuts, []},
          {Credo.Check.Refactor.MapMap, []},
          {Credo.Check.Refactor.ModuleDependencies, []},
          {Credo.Check.Refactor.NegatedIsNil, []},
          {Credo.Check.Refactor.PassAsyncInTestCases, []},
          {Credo.Check.Refactor.PipeChainStart, []},
          {Credo.Check.Refactor.RejectFilter, []},
          {Credo.Check.Refactor.VariableRebinding, []},
          {Credo.Check.Warning.LazyLogging, []},
          {Credo.Check.Warning.LeakyEnvironment, []},
          {Credo.Check.Warning.MapGetUnsafePass, []},
          {Credo.Check.Warning.MixEnv, []},
          {Credo.Check.Warning.UnsafeToAtom, []},
          {Credo.Check.Refactor.MapInto, []}

          #
          # Custom checks can be created using `mix credo.gen.check`.
          #
        ]
      }
    }
  ]
}


lib/cli/cli.ex
defmodule Commandline.CLI do
  @moduledoc """
  Command line interface for checking payment card account number brands.

  Build using mix escript.build from project root.

  ## Examples

      iex> ./simplecardbrand 41111111111111111
      PAN: 41111111111111111 -> Unknown card brand.

      iex> ./simplecardbrand 4111111111111111 7111111111111111 8331111111111111111
      PAN: 4111111111111111 -> Unknown card brand.
      PAN: 7111111111111111 -> Brand: gpn
      PAN: 8331111111111111111 -> Unknown card brand.

      iex> ./simplecardbrand 4111111 --pan-length=16
      PAN: 4111111 -> Brand: visa
  """

  # https://inquisitivedeveloper.com/learn-with-me-elixir-elixirlargesort-intgen-project-part-2-77/
  # https://clairettran.medium.com/elixir-beginner-ii-tutorial-ab219ba6f4cd
  @doc """
  Entry point.

  See module documentation for usage.
  """
  def main(args) do
    {opts, argv, _errors} =
      OptionParser.parse(args,
        switches: [pan_length: :integer],
        aliases: []
      )

    opts
    |> Enum.into(%{})
    |> _brandit(argv)
  end

  defp _brandit(%{pan_length: pan_length} = flags, [pan | pans]) do
    case SimpleCardBrand.card_brand(pan, pan_length) do
      {:ok, brand} -> IO.puts("PAN: #{pan} -> Brand: #{brand}")
      {:error, {_, reason}} -> IO.puts("PAN: #{pan} -> #{reason}")
    end

    _brandit(flags, pans)
  end

  defp _brandit(%{}, [pan | pans]) do
    case SimpleCardBrand.card_brand(pan) do
      {:ok, brand} -> IO.puts("PAN: #{pan} -> Brand: #{brand}")
      {:error, {_, reason}} -> IO.puts("PAN: #{pan} -> #{reason}")
    end

    _brandit(%{}, pans)
  end

  defp _brandit(%{}, []) do
  end
end


lib/simple_card_brand.ex
defmodule SimpleCardBrand.Guards do
  @moduledoc """
  Guard macros.

  NOTE: defguard is a macro so the doc tests do not run.

  """
  @moduledoc since: "0.1.1"

  @doc """
  Check the PAN length is in the minimum to maximum allowed length (inclusive).
  Use in guards where `pan_length` is known to be an integer.

  ## Examples

      iex> SimpleCardBrand.Guards.pan_range(16, 12, 19)
      true

      iex> SimpleCardBrand.Guards.pan_range(16, 19, 19)
      false

  """
  defguard pan_range(pan_length, minimum, maximum) when pan_length in minimum..maximum

  @doc """
  Compare the first eight (8) digits of the `pan` against the UkrCard range.
  Use in guards where `pan` is known to be bytes.

  ## Examples

      iex> SimpleCardBrand.Guards.is_ukrcard("60420099123212")
      true

      iex> SimpleCardBrand.Guards.is_ukrcard("6042010022222")
      false
  """
  defguard is_ukrcard(pan)
           when binary_part(pan, 0, 8) >= "60400100" and binary_part(pan, 0, 8) <= "60420099"

  @doc """
  Compare the first six (6) digits of the `pan` against the Verve brand range.
  Use in guards where `pan` is known to be bytes.

  ## Examples

      iex> SimpleCardBrand.Guards.is_verve("506099212121")
      true

      iex> SimpleCardBrand.Guards.is_verve("50609123")
      false
  """
  defguard is_verve(pan)
           when (binary_part(pan, 0, 6) >= "506099" and binary_part(pan, 0, 6) <= "506198") or
                  (binary_part(pan, 0, 6) >= "507865" and binary_part(pan, 0, 6) <= "507964") or
                  (binary_part(pan, 0, 6) >= "650002" and binary_part(pan, 0, 6) <= "650027")
end

defmodule SimpleCardBrand do
  @moduledoc """
  Identify the card brand from the PAN or the first six or  eight (UkrCard) PAN digits.
  The PAN must contain only digits without leading or trailing spaces.

  Supports:
  - American Express (:americanexpress)
  - BORICA (:borica)
  - China T-Union (:chinatunion)
  - China UnionPay (:chinaunionpay)
  - Dankort (:dankort)
  - Diners Club (:dinersclub)
  - Diners Club International (:dinersclubinternational)
  - Discover (:discover)
  - GPN (:gpn)
  - Humo (:humo)
  - InstaPayment (:instapayment)
  - InterPayment (:interpayment)
  - JCB: (:jcb)
  - LankaPay: (:lankapay)
  - Maestro (:maestro)
  - Maestro UK (:maestrouk)
  - Mastercard (:mastercard)
  - Mir (:mir)
  - Napas (:napas)
  - RuPay (:rupay)
  - Troy (:troy)
  - UATP (:uatp)
  - UkrCard (:ukrcard)
  - UzCard (:uzcard)
  - Verve (:verve)
  - Visa (:visa)
  - Visa Electron (:visaelectron)

  Rules as per: https://en.wikipedia.org/wiki/Payment_card_number

  """
  @moduledoc since: "0.1.1"

  import SimpleCardBrand.Guards

  @minimum_pan_length 12
  @maximum_pan_length 19

  @spec card_brand(String.t()) ::
          {:error,
           {:pan_too_long, String.t()}
           | {:pan_too_short, String.t()}
           | {:pan_unknown, String.t()}}
          | {:ok, atom}
  @doc ~S"""
  Identify the card brand from the `pan`.

  ## Examples

      iex> SimpleCardBrand.card_brand("4111111111111111")
      {:ok, :visa}

      iex> SimpleCardBrand.card_brand("6040010012121161819")
      {:ok, :ukrcard}

      iex> SimpleCardBrand.card_brand("5060994444444416")
      {:ok, :verve}

      iex> SimpleCardBrand.card_brand("41111111111")
      {:error, {:pan_too_short, "Minimum PAN length is 12, found 11."}}

      iex> SimpleCardBrand.card_brand("41111111111111111120")
      {:error, {:pan_too_long, "Maximum PAN length is 19, found 20."}}

  """
  def card_brand(pan) when is_binary(pan) do
    pan
    |> card_brand(String.length(pan))
  end

  @spec card_brand(String.t(), integer) ::
          {:error,
           {:pan_too_long, String.t()}
           | {:pan_too_short, String.t()}
           | {:pan_unknown, String.t()}}
          | {:ok, atom}
  @doc ~S"""
  Identify the card brand from a full or partial `pan` and the actual PAN length.
  Useful when identifying brands from previously-stored PANs.

  Check for UkrCard and Verve early.

  ## Examples

      iex> SimpleCardBrand.card_brand("411111", 16)
      {:ok, :visa}

      iex> SimpleCardBrand.card_brand("6040010012", 18)
      {:ok, :ukrcard}

      iex> SimpleCardBrand.card_brand("50609944444", 19)
      {:ok, :verve}

      iex> SimpleCardBrand.card_brand("4111111111111111", 10)
      {:error, {:pan_too_short,"Minimum PAN length is 12, found 10."}}

      iex> SimpleCardBrand.card_brand("0123456789012345678")
      {:error, {:pan_unknown,"Unknown card brand."}}
  """
  def card_brand(pan, pan_length)
      when is_binary(pan) and is_integer(pan_length) and pan_length in [16, 18, 19] and
             is_verve(pan) do
    {:ok, :verve}
  end

  def card_brand(pan, pan_length)
      when is_binary(pan) and is_integer(pan_length) and pan_length in [16, 18, 19] and
             is_ukrcard(pan) do
    {:ok, :ukrcard}
  end

  def card_brand(pan, pan_length) when is_binary(pan) and is_integer(pan_length) do
    cond do
      pan_length < @minimum_pan_length ->
        {:error,
         {:pan_too_short,
          "Minimum PAN length is " <>
            Integer.to_string(@minimum_pan_length) <>
            ", found " <> Integer.to_string(pan_length) <> "."}}

      pan_length > @maximum_pan_length ->
        {:error,
         {:pan_too_long,
          "Maximum PAN length is " <>
            Integer.to_string(@maximum_pan_length) <>
            ", found " <> Integer.to_string(pan_length) <> "."}}

      true ->
        _card_brand(String.codepoints(pan), pan_length)
    end
  end

  # China T-Union
  defp _card_brand(["3", "1" | _], 19) do
    {:ok, :chinatunion}
  end

  # American Express
  defp _card_brand(["3", second | _], 15) when second in ["4", "7"] do
    {:ok, :americanexpress}
  end

  # LankaPay
  defp _card_brand(["3", "5", "7", "1", "1", "1" | _], 16) do
    {:ok, :lankapay}
  end

  # RuPay
  defp _card_brand(["3", "5", "3" | _], 16) do
    {:ok, :rupay}
  end

  # RuPay
  defp _card_brand(["3", "5", "6" | _], 16) do
    {:ok, :rupay}
  end

  # JCB
  defp _card_brand(["3", "5" | pan], pan_length) when pan_range(pan_length, 16, 19) do
    sub_pan =
      Enum.slice(pan, 0, 2)
      |> Enum.join()

    if "28" <= sub_pan and sub_pan <= "89" do
      {:ok, :jcb}
    else
      {:error, {:pan_unknown, "Unknown card brand."}}
    end
  end

  # Diners Club International
  defp _card_brand(["3", "6" | _], pan_length) when pan_range(pan_length, 14, 19) do
    {:ok, :dinersclubinternational}
  end

  # Diners Club International
  # 54 is in the Mastercard range. Branded as Diners in US and Canada
  defp _card_brand(["5", "4" | _], 16) do
    {:ok, :dinersclub}
  end

  # BORICA
  defp _card_brand(["2", "2", "0", "5" | _], 16) do
    {:ok, :borica}
  end

  # Mir
  defp _card_brand(["2", "2", "0", fourth | _], pan_length)
       when fourth in ["0", "1", "2", "3", "4"] and pan_range(pan_length, 16, 19) do
    {:ok, :mir}
  end

  # Mastercard and GPN starting with 2
  defp _card_brand(["2" | pan], 16) do
    sub_pan =
      Enum.slice(pan, 0, 3)
      |> Enum.join()

    if "221" <= sub_pan and sub_pan <= "720" do
      {:ok, :mastercard}
    else
      {:ok, :gpn}
    end
  end

  # Mastercard
  # 54 cards are branded as Diners in US and Canada.
  defp _card_brand(["5", second | _], 16) when second in ["1", "2", "3", "5"] do
    {:ok, :mastercard}
  end

  # Visa Electron
  defp _card_brand(["4", "0", "2", "6" | _], 16) do
    {:ok, :visaelectron}
  end

  # Visa Electron
  defp _card_brand(["4", "1", "7", "5", "0", "0" | _], 16) do
    {:ok, :visaelectron}
  end

  # Visa Electron
  defp _card_brand(["4", "5", "0", "8" | _], 16) do
    {:ok, :visaelectron}
  end

  # Visa Electron
  defp _card_brand(["4", "8", "4", "4" | _], 16) do
    {:ok, :visaelectron}
  end

  # Visa Electron
  defp _card_brand(["4", "9", "1", fourth | _], 16) when fourth in ["3", "7"] do
    {:ok, :visaelectron}
  end

  # Visa
  defp _card_brand(["4" | _], pan_length) when pan_length in [13, 16, 19] do
    {:ok, :visa}
  end

  # Discover
  defp _card_brand(["6", "0", "1", "1" | _], pan_length) when pan_range(pan_length, 16, 19) do
    {:ok, :discover}
  end

  # Discover
  defp _card_brand(["6", "4", third | _], pan_length)
       when third in ["4", "5", "6", "7", "8", "9"] and pan_range(pan_length, 16, 19) do
    {:ok, :discover}
  end

  # Discover
  defp _card_brand(["6", "2", "2" | tail], pan_length) when pan_range(pan_length, 16, 19) do
    sub_pan =
      Enum.slice(tail, 0, 3)
      |> Enum.join()

    if "126" <= sub_pan and sub_pan <= "925" do
      {:ok, :discover}
    else
      # 62 is ChinaUnionPay
      # Both Discover and ChinaUnionPay length can be 16 to 19.
      {:ok, :chinaunionpay}
    end
  end

  # Discover
  defp _card_brand(["6", "5" | _], pan_length) when pan_range(pan_length, 16, 19) do
    {:ok, :discover}
  end

  # China UnionPay: After Discover 622
  defp _card_brand(["6", "2" | _], pan_length) when pan_range(pan_length, 16, 19) do
    {:ok, :chinaunionpay}
  end

  # Maestro UK
  # Conflicts with Maestro range.
  defp _card_brand(["6", "7", "5", "9" | _], pan_length) when pan_range(pan_length, 12, 19) do
    {:ok, :maestrouk}
  end

  # Maestro UK
  defp _card_brand(["6", "7", "6", "7", "7", sixth | _], pan_length)
       when sixth in ["0", "4"] and pan_range(pan_length, 12, 19) do
    {:ok, :maestrouk}
  end

  # Maestro
  defp _card_brand(["5", "0", "1", "8" | _], pan_length) when pan_range(pan_length, 12, 19) do
    {:ok, :maestro}
  end

  # Maestro
  defp _card_brand(["5", "0", "2", "0" | _], pan_length) when pan_range(pan_length, 12, 19) do
    {:ok, :maestro}
  end

  # Maestro
  defp _card_brand(["5", "0", "3", "8" | _], pan_length) when pan_range(pan_length, 12, 19) do
    {:ok, :maestro}
  end

  # Maestro
  defp _card_brand(["5", "8", "9", "3" | _], pan_length) when pan_range(pan_length, 12, 19) do
    {:ok, :maestro}
  end

  # Maestro
  defp _card_brand(["6", "3", "0", "4" | _], pan_length) when pan_range(pan_length, 12, 19) do
    {:ok, :maestro}
  end

  # Maestro
  defp _card_brand(["6", "7", "6", fourth | _], pan_length)
       when fourth in ["1", "2", "3"] and pan_range(pan_length, 12, 19) do
    {:ok, :maestro}
  end

  # UATP
  defp _card_brand(["1" | _], 15) do
    {:ok, :uatp}
  end

  # Dankort
  defp _card_brand(["5", "0", "1", "9" | _], 16) do
    {:ok, :dankort}
  end

  # RuPay
  defp _card_brand(["6", "0" | _], 16) do
    {:ok, :rupay}
  end

  # RuPay
  defp _card_brand(["8", "1" | _], 16) do
    {:ok, :rupay}
  end

  # RuPay
  defp _card_brand(["8", "2" | _], 16) do
    {:ok, :rupay}
  end

  # RuPay
  defp _card_brand(["5", "0", "8" | _], 16) do
    {:ok, :rupay}
  end

  # InstaPayment
  defp _card_brand(["6", "3", third | _], pan_length)
       when third in ["7", "8", "9"] and pan_range(pan_length, 16, 19) do
    {:ok, :instapayment}
  end

  # InterPayment
  defp _card_brand(["6", "3", "6" | _], pan_length) when pan_range(pan_length, 16, 19) do
    {:ok, :interpayment}
  end

  # UzCard
  defp _card_brand(["8", "6", "0", "0" | _], 16) do
    {:ok, :uzcard}
  end

  # Napas
  defp _card_brand(["9", "7", "0", "4" | _], 16) do
    {:ok, :napas}
  end

  # Napas
  defp _card_brand(["9", "7", "9", "2" | _], 16) do
    {:ok, :troy}
  end

  # Humo
  defp _card_brand(["9", "8", "6", "0" | _], 16) do
    {:ok, :humo}
  end

  # GPN starting with 2 handled by Mastercard pattern match.
  defp _card_brand([first | _], 16) when first in ["1", "6", "7", "8", "9"] do
    {:ok, :gpn}
  end

  # Error
  defp _card_brand(_, _) do
    {:error, {:pan_unknown, "Unknown card brand."}}
  end
end
